pyomo/core/base/indexed_component_slice.py

```diff
@@ -420,33 +449,80 @@ def next(self):
 
     def __next__(self):
         """Return the next element in the slice."""
+        # In each call to this function, idx will initially point
+        # to the top of the stack.
         idx = len(self._iter_stack)-1
         while True:
             # Flush out any non-slice levels.  Since we initialize
             # _iter_stack with None, in the first call this will
-            # immediately walk up to the beginning of the _iter_stack
+            # immediately walk down to the beginning of the _iter_stack
+            #
+            # On subsequent calls, we walk down to the nearest active
+            # iterator, which is left over if the last call to this
+            # function encountered a slice. Previous iterators could
+            # be active as well, but we wait to deal with them until
+            # we have (temporarily) exhausted the "highest level"
+            # iterator.
             while self._iter_stack[idx] is None:
                 idx -= 1
             # Get the next element in the deepest active slice
             try:
                 if self._iter_stack[idx] is _NotIterable:
+                    # When does this happen?
```

Comment:
Am I correct to assert that `_ReferenceDict.__setitem__` only gets hit if we are directly working with a `_ReferenceDict` object directly, rather than with the reference itself? E.g. 
```python
ref = Reference(m.x[:])
ref[1] = 10
```
would not hit `_ReferenceDict.__setitem__` and therefore would not hit the `_NotIterable` code we are talking about.

Reply:
Yes.  I believe that is correct.